rules/validator/validator.go

```diff
@@ -85,147 +97,299 @@ func (v *validator) validateNamespace(ns string) error {
 		if err != nil {
-			return err
+			return fmt.Errorf("mapping rule '%s' has invalid filter %s: %v", rule.Name, rule.Filter, err)
 		}
 
 		// Validate the metric types.
 		types, err := v.opts.MetricTypesFn()(filterValues)
 		if err != nil {
-			return err
+			return fmt.Errorf("mapping rule '%s' cannot infer metric types from filter %v: %v", rule.Name, rule.Filter, err)
 		}
 		if len(types) == 0 {
-			return fmt.Errorf("mapping rule '%s' does not match any allowed metric types, filter=%s", view.Name, view.Filter)
+			return fmt.Errorf("mapping rule '%s' does not match any allowed metric types, filter=%s", rule.Name, rule.Filter)
 		}
 
-		// Validate that the policies are valid.
-		if err := v.validatePolicies(view.Name, view.Policies, types); err != nil {
-			return err
+		// Validate the aggregation ID.
+		if err := v.validateAggregationID(rule.AggregationID, firstLevelAggregationType, types); err != nil {
+			return fmt.Errorf("mapping rule '%s' has invalid aggregation id %v: %v", rule.Name, rule.AggregationID, err)
+		}
+
+		// Validate that the storage policies are valid.
+		if err := v.validateStoragePolicies(rule.StoragePolicies, types); err != nil {
+			return fmt.Errorf("mapping rule '%s' has invalid storage policies in %v: %v", rule.Name, rule.StoragePolicies, err)
 		}
 	}
 	return nil
 }
 
 func (v *validator) validateRollupRules(rrv map[string]*models.RollupRuleView) error {
 	var (
-		namesSeen   = make(map[string]struct{}, len(rrv))
-		targetsSeen = make([]models.RollupTargetView, 0, len(rrv))
+		namesSeen = make(map[string]struct{}, len(rrv))
+		pipelines = make([]mpipeline.Pipeline, 0, len(rrv))
 	)
-	for _, view := range rrv {
+	for _, rule := range rrv {
+		if rule.Tombstoned {
+			continue
+		}
 		// Validate that no rules with the same name exist.
-		if _, exists := namesSeen[view.Name]; exists {
-			return errors.NewRuleConflictError(fmt.Sprintf("rollup rule '%s' already exists", view.Name))
+		if _, exists := namesSeen[rule.Name]; exists {
+			return merrors.NewRuleConflictError(fmt.Sprintf("rollup rule '%s' already exists", rule.Name))
 		}
-		namesSeen[view.Name] = struct{}{}
+		namesSeen[rule.Name] = struct{}{}
 
 		// Validate that the filter is valid.
-		filterValues, err := v.validateFilter(view.Name, view.Filter)
+		filterValues, err := v.validateFilter(rule.Filter)
 		if err != nil {
-			return err
+			return fmt.Errorf("rollup rule '%s' has invalid filter %s: %v", rule.Name, rule.Filter, err)
 		}
 
 		// Validate the metric types.
 		types, err := v.opts.MetricTypesFn()(filterValues)
 		if err != nil {
-			return err
+			return fmt.Errorf("rollup rule '%s' cannot infer metric types from filter %v: %v", rule.Name, rule.Filter, err)
 		}
 		if len(types) == 0 {
-			return fmt.Errorf("rollup rule '%s' does not match any allowed metric types, filter=%s", view.Name, view.Filter)
+			return fmt.Errorf("rollup rule '%s' does not match any allowed metric types, filter=%s", rule.Name, rule.Filter)
```

Comment:
So my thoughts on this is that `MetricTypesFn()` is the function that infers what possible metric types this rule may apply to. If the *inference* logic fails, the function should return an error. However if it succeeds (even though the list of types inferred is an empty list), then as far as the `MetricTypesFn` is concerned, it's done its job to infer the metric types and should return no error in this case. 

Also we do support cases when `len(types) > 1`. In fact if a rule doesn't specific the metric type filter, or does `type:*`, then it can match multiple metric types.

Reply:
I see, then the rollup tags will enforce metrics from different tags won't be rollup together? Then this should be fine

Reply:
Yeah so `type` is a required rollup tag, so users don't end up rolling counters and gauges into one rollup metric.